matlab/HyEQsolver_new_slow.m

```diff
@@ -0,0 +1,434 @@
+                    % solver, then larger steps sizes won't improve our
+                    % solution and might cause use to miss the desired
+                    % step, so we simply use the ODE solution.
+                    t_next = t_flow(2);
+                    x_next = x_flow(2,:)';
+                    break;
+                end
+                x_now = x_total(index_last, :)';
+                t_next = t_total(index_last) + delta_t;
+                x_next = x_now + delta_t * f(x_now,t_total(index_last),j);
+                next_in_C = C(x_next,t_next,j);
+                next_in_D = D(x_next,t_next,j);
+                next_will_not_flow = ~next_in_C || (rule == 1 && next_in_D);
+                if next_will_not_flow
+                    break;
+                end
+            end
+            
+            if (index_last + 1 > size(t_total, 1))
+                x_total = [x_total; zeros(1, cx)];
+                t_total = [t_total; 0];
+                j_total = [j_total; 0];
+            end     
+
+            x_total(index_last + 1 : index_last + nt, :) = x_flow;
+            t_total(index_last + 1 : index_last + nt, :) = t_flow;
+            j_total(index_last + 1 : index_last + nt, :) = j*ones(nt, 1);
+            index_last = index_last + nt;
+
+        end
+    elseif doJump
+        [j, t_total, j_total, x_total, index_last] = jump(g,j, ...
+         t_total,j_total,x_total,index_last, prealloc_size);
+        progress.setJ(j);
+    end
+end
+
+t = t_total(1 : index_last, :);
+x = x_total(1 : index_last, :);
+j = j_total(1 : index_last, :);
+
+progress.done();
+
+end
+
+function [value,isterminal,direction] = zeroevents(x,t,j,C,D,rule)
+% ZEROEVENTS Creates an event to terminate the ode solver when the state leaves C or enters D (if rule == 1).
+
+insideC = C(x,t,j);
+insideD = D(x,t,j);
+switch rule
+    case 1 % -> priority for jumps
+        % For jump priority, we terminate flows whenever (1) the solution 
+        % leaves C, or (2) the solution enters D. 
+        stop = ~insideC || insideD;
+    case 2 % -> priority for flows
+        % For flow priority, we terminate flows whenever the solution 
+        % leaves C.
+        stop = ~insideC;
+end
+if any(isnan(x)) || any(isinf(x))
+    stop = 1;
+end
+isterminal = 1;
+value = 1 - stop;
+direction = -1;
+
+end
+
+function [j,tout,jout,xout,index_last] = jump(g,j,tout,jout,xout,index_last,prealloc_size)
+% Jump
+j = j+1;
+n = size(xout, 2);
+y = g(xout(index_last,:).',tout(index_last),jout(index_last)); 
+assert(size(y, 1) == n, 'Output of jump map was expected to be %d but instead was %d', n, size(y, 1))
+% Save results
+
+if (index_last + 1 > size(tout, 1))
+    xout = [xout; zeros(prealloc_size, size(xout, 2))];
+    tout = [tout; zeros(prealloc_size, 1)];
+    jout = [jout; zeros(prealloc_size, 1)];
+end
+
+tout(index_last + 1, 1) = tout(index_last, 1);
+xout(index_last + 1, :) = y';
+jout(index_last + 1, 1) = j;
+index_last = index_last  + 1;
```

Comment:
Is line 408 here where the slowdown occurs? Does `xout` have the same size as `tout`? Maybe we are unexpectedly increasing the array size.